File: 프로그래머스/lv2/60057. 문자열 압축/문자열 압축.py
def solution(s):
    answer = len(s)
    
    for i in range(1, len(s) // 2 + 1):
        tmp = ''
        cnt = 1
        token = s[:i]
        
        for j in range(i, len(s), i):
            if token == s[j:i + j]:
                cnt += 1
            else:
                if cnt != 1:
                    tmp += str(cnt) + token
                else:
                    tmp += token
                token = s[j:j + i]
                cnt = 1
        if cnt != 1:
            tmp += str(cnt) + token
        else:
            tmp += token
            
        answer = min(answer, len(tmp))
    
    return answer

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.java
import java.util.*;

class Solution {
    
    int answer;
    
    public int solution(String s) {
        answer = s.length();
        
        for (int i = 1; i <= s.length() / 2; i++) {
            check(s, i);
        }
        
        return answer;
    }
    
    public void check(String s, int unitLength) {
        List<String> div = new ArrayList<>();
        int idx = 0;
        
        while (idx < s.length()) {
            if (idx + unitLength >= s.length()) {
                div.add(s.substring(idx));
                break;
            }
            
            String newS = s.substring(idx, idx + unitLength);
            div.add(newS);
            
            idx += unitLength;
        }
        
        int tmp = 0;
        boolean[] visited = new boolean[div.size()];
        
        for (int i = 0; i < div.size(); i++) {
            if (visited[i]) continue;
            
            int cnt = 1;
            String now = div.get(i);
            for (int j = i + 1; j < div.size(); j++) {
                if (now.equals(div.get(j))) {
                    visited[j] = true;
                    cnt += 1;
                }
                
                else {
                    break;
                }
            }
            
            tmp += now.length();
            
            if (cnt > 1) {
                tmp += String.valueOf(cnt).length();
            }
        }
        
        answer = Math.min(answer, tmp);
    }
}